### PR TITLE
middleware: add wasm basic

### PR DESCRIPTION
**Pull request by daixiang0**
Signed-off-by: Loong <loong.dai@intel.com>

# Description

Basic WASM HTTP middleware implement.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: https://github.com/dapr/components-contrib/issues/1171
## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by daixiang0** [comment]
@shubham1172 could you help check the exact error of fossa? And where is the webapp:
> The build failed. Check the FOSSA webapp for more details.

**Comment by shubham1172** [comment]
@daixiang0 you can check out the Dapr project once you login to https://app.fossa.com/projects
As of the build failure, it looks like a FOSSA problem. Scan completed fine; the test step failed without any other information. I don't see anything on the FOSSA portal too. Had created this issue to track https://github.com/fossas/fossa-cli/issues/943

**Comment by daixiang0** [comment]
@yaron2 could you take a look?

**Comment by yaron2** [comment]
@daixiang0 please resolve the conflict.

**Comment by yaron2** [comment]
ping @daixiang0

**Comment by yaron2** [comment]
Can you give an example here of how a component YAML would look like for WASM?

File: middleware/http/wasm/basic/basic.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+			} else {
+				ctx.Request.SetRequestURIBytes(bytes)
+			}
+
+			h(ctx)
+		}
+	}, nil
+}
+
+func (m *Middleware) getNativeetadata(metadata middleware.Metadata) (*middlewareMetadata, error) {
```
Comment (berndverst) [inline middleware/http/wasm/basic/basic.go:126]:
```suggestion
func (m *Middleware) getNativeMetadata(metadata middleware.Metadata) (*middlewareMetadata, error) {
```

File: middleware/http/wasm/basic/basic.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	return &Middleware{logger: logger}
+}
+
+// GetHandler returns the HTTP handler provided by wasm basic middleware.
+func (m *Middleware) GetHandler(metadata middleware.Metadata) (func(h fasthttp.RequestHandler) fasthttp.RequestHandler, error) {
+	var (
+		meta *middlewareMetadata
+		err  error
+	)
+	meta, err = m.getNativeetadata(metadata)
```
Comment (berndverst) [inline middleware/http/wasm/basic/basic.go:39]:
```suggestion
	meta, err = m.getNativeMetadata(metadata)
```

**Review by berndverst — CHANGES_REQUESTED** [review]

**Comment by berndverst** [comment]
@yaron2 component should look like this if I'm understanding this PR correctly

```
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: basicwasm
spec:
  type: middleware.http.wasm.basic
  version: v1
  metadata:
  - name: path
    value: "./hello.wasm"
  - bane: runtime
    value: "wazero"
```

**Comment by daixiang0** [comment]
> @yaron2 component should look like this if I'm understanding this PR correctly
> 
> ```
> apiVersion: dapr.io/v1alpha1
> kind: Component
> metadata:
>   name: basicwasm
> spec:
>   type: middleware.http.wasm.basic
>   version: v1
>   metadata:
>   - name: path
>     value: "./hello.wasm"
>   - bane: runtime
>     value: "wazero"
> ```

@berndverst totally correct!

**Review by berndverst — APPROVED** [review]

**Review by yaron2 — APPROVED** [review]

**Comment by ewassef** [comment]
What is the WASM interface or structure? any guidance here? Also, can this be used with Pub/Sub or just HTTP service invocation?

**Comment by codefromthecrypt** [comment]
@ewassef sorry about the large delay in responding. I'm just reviewing old issues for the dapr talk at kubecon EU next week.

> What is the WASM interface or structure? any guidance here? Also, can this be used with Pub/Sub or just HTTP service invocation?

Since this PR landed, we had requests to make the model more flexible. We wrote a specification and some SDKs for [http-wasm](https://http-wasm.io/), though currently this is only used by dapr (PR for mosn has been paused due to resources)

So, as of v1.10 this feature is more fleshed out and [documented](https://docs.dapr.io/reference/components-reference/supported-middleware/middleware-wasm/) and has a [sample](https://github.com/dapr/samples/tree/master/hello-wasm), also.

About Pub/Sub, the closest we have in progress now is the output binding, which is incomplete, but I plan to help with this after KubeCon EU next week. It is pure WASI, so there's no pre-defined model (basically you interact via stdin or CLI args). You can subscribe to [this issue](https://github.com/dapr/components-contrib/issues/2274) which I think is prematurely closed as we haven't documented or integrated it into dapr, yet.

**Pull request merged**